src/ctap2/attestation.rs

```diff
@@ -270,23 +220,39 @@ impl<'de> Deserialize<'de> for AuthenticatorData {
                 formatter.write_str("a byte array")
             }
 
-            fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+            fn visit_bytes<E>(self, input: &[u8]) -> Result<Self::Value, E>
             where
                 E: SerdeError,
             {
-                parse_ad(v)
-                    .map(|(_input, value)| value)
-                    .map_err(|e| match e {
-                        NomErr::Incomplete(nom::Needed::Size(len)) => {
-                            E::invalid_length(v.len(), &format!("{}", v.len() + len.get()).as_ref())
-                        }
-                        NomErr::Incomplete(nom::Needed::Unknown) => {
-                            E::invalid_length(v.len(), &"unknown") // We don't know the expected value
-                        }
-                        // TODO(baloo): is that enough? should we be more
-                        //              specific on the error type?
-                        e => E::custom(e.to_string()),
-                    })
+                let mut cursor = Cursor::new(input);
+                let mut rp_id_hash_raw = [0u8; 32];
+                cursor
+                    .read_exact(&mut rp_id_hash_raw)
+                    .map_err(|_| serde_parse_err("32 bytes"))?;
+                let rp_id_hash = RpIdHash(rp_id_hash_raw);
+
+                // be conservative, there is a couple of reserved values in
+                // AuthenticatorDataFlags, just truncate the one we don't know
+                let flags = AuthenticatorDataFlags::from_bits_truncate(read_byte(&mut cursor)?);
+                let counter = read_be_u32(&mut cursor)?;
+                let mut credential_data = None;
+                if flags.contains(AuthenticatorDataFlags::ATTESTED) {
+                    credential_data = Some(parse_attested_cred_data(&mut cursor)?);
+                }
+
+                let mut extensions = None;
+                if flags.contains(AuthenticatorDataFlags::EXTENSION_DATA) {
+                    extensions = Some(from_slice_stream(&mut cursor)?);
+                }
```

Comment:
You don't have to wrap with `Option` if you add an `else { Default::default() }` case.

Reply:
Ah, yes! That's much nicer.